Create migration, models and schema for student identifier

dinever opened the pull request:
This is a part of #614.

tarebyte commented on app/models/assignment_invitation.rb:
```diff
@@ -24,6 +24,14 @@ def redeem_for(invitee)
+                                        student_identifier_type: assignment.student_identifier_type,
+                                        value: identifier_value)
+  end
+
```
I think we can ✂️ this bit for this PR.

tarebyte commented:
@dinever can we bring the model tests into this PR as well?

dinever commented:
Hi @tarebyte, 

Just added some tests for the models in 600cf9f62c4ec45fbafe9eb6121c35eeb5b31709. Haven't written any model tests before so I didn't come up with many test cases. So I wonder if there is anything else that we need to test about the models. 😄 

Thanks!

tarebyte commented on app/models/student_identifier_type.rb:
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+class StudentIdentifierType < ActiveRecord::Base
+  default_scope { where(deleted_at: nil) }
```
I think we can ✂️ this.

dinever replied:
Hi @tarebyte, I wonder the reason why we need to ✂️ this. This line seems to be necessary for soft delete. It would be great if you can elaborate a little bit on this. Thanks!

tarebyte replied:
@dinever I forgot that at some point we will be querying this. We can leave it.

tarebyte commented on app/models/student_identifier.rb:
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+class StudentIdentifier < ActiveRecord::Base
+  default_scope { where(deleted_at: nil) }
```
Same [comment](https://github.com/education/classroom/pull/648/files#r68456823) here

tarebyte commented on spec/support/github_factory.rb:
```diff
@@ -12,4 +12,11 @@ def self.create_owner_classroom_org
+                                  name: 'Test',
+                                  description: 'Test',
+                                  content_type: 'text')
+  end
```
I'd rather we use FactoryGirl, and just specify the organization.

dinever commented:
Hi @tarebyte @johndbritton, I think this one is ready for a code review. Could you please take a 👀  at it when you're available?

tarebyte commented on db/migrate/20160524234655_create_student_identifiers.rb:
```diff
@@ -0,0 +1,22 @@
+      t.string :name, null: false
+      t.string :description, null: false
+      t.integer :content_type, null: false
+      t.datetime :deleted_at
```
We should probably have the `created_at` and `updated_at` columns here as well

tarebyte commented on db/migrate/20160524234655_create_student_identifiers.rb:
```diff
@@ -0,0 +1,22 @@
+      t.belongs_to :user,         index: true
+      t.belongs_to :student_identifier_type
+      t.string :value, null: false
+      t.datetime :deleted_at
```
Same [comment](https://github.com/education/classroom/pull/648/files#r69911304) here

tarebyte commented on db/migrate/20160524234655_create_student_identifiers.rb:
```diff
@@ -0,0 +1,22 @@
+    create_table :student_identifiers do |t|
+      t.belongs_to :organization, index: true
+      t.belongs_to :user,         index: true
+      t.belongs_to :student_identifier_type
```
This should also probably be indexed, thoughts?

tarebyte commented on spec/models/student_identifier_spec.rb:
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+RSpec.describe StudentIdentifier, type: :model do
+  include ActiveJob::TestHelper
```
Do we need this line?

tarebyte commented on spec/models/student_identifier_spec.rb:
```diff
@@ -0,0 +1,32 @@
+    let(:user)                    { organization.users.first }
+
+    let(:student_identifier_type) { create(:student_identifier_type, organization: organization)                     }
+    let(:group_assignment)        { create(:group_assignment, organization: organization, user: user, value: 'Test') }
```
`group_assignment` doesn't have a value column. Copy paste error?

tarebyte commented on spec/models/student_identifier_spec.rb:
```diff
@@ -0,0 +1,32 @@
+             student_identifier_type: student_identifier_type,
+             value: 'Test')
+
+      new_group_assignment = build(:student_identifier,
```
Same question https://github.com/education/classroom/pull/648/files#r69911841 here

tarebyte commented on spec/models/student_identifier_type_spec.rb:
```diff
@@ -0,0 +1,8 @@
+require 'rails_helper'
+
+RSpec.describe StudentIdentifierType, type: :model do
+  include ActiveJob::TestHelper
```
Same question https://github.com/education/classroom/pull/648/files#r69911555 here

dinever commented:
Hi @tarebyte, all changes mentioned above have been done. Thanks a lot!

Changes:

- Added `updated_at` and `created_at` fields on `StudentIdentifiers` and `StudentIdentifierTypes`: 71a65ff4f74f76b49e0cba32aa44403ac9da9872
- Indexed `student_identifier_types` for `student_identifiers`: 18fd48ce4ca300e5e1a691a91ef6cd520c825932
- Fixed naming mistake: 26d10c91cce80b534028b166d0ee99e6dcba1ed6
- Removed unused lines: 2266db51cc56bf80ec37f9ec24ce89a465ac3e1c and b79b37b1e830377e3e1abfbab319b867a858de64
- Removed this method, which will be added in a following PR: 06529ed20b5213934c5765b27293c737c489b1bf

tarebyte commented:
Migration successful 🎉
